core/src/main/java/org/infinispan/persistence/sifs/Compactor.java

```diff
@@ -402,9 +399,16 @@ public void accept(Object o) throws Throwable {
                      continue;
                   }
                   // Make sure we don't start another compaction for this file while performing expiration
-                  stats.setScheduled();
+                  if (stats.setScheduled()) {
+                     compactSingleFile(fileId, isLogFile, subscriber, currentTimeMilliseconds);
+                     if (isLogFile) {
+                        // Unschedule the compaction for log file as we can't remove it
+                        stats.scheduled.set(false);
+                     }
+                  }
+               } else {
+                  log.tracef("Skipping expiration for file %d as it is not included in fileStats", fileId);
                }
-               compactSingleFile(fileId, isLogFile, subscriber, currentTimeMilliseconds);
```

Comment:
I have no idea why this was ever allowed to run compaction without any stat for the file. This if statement was somehow true in some rare cases which would cause the double compaction of the same file if it wasn't deleted between.